File: array/problem121/problem121.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        // One pass: keep cheapest buy price and best sell profit
        int minPrice = prices[0];
        int best = 0;
        for (int i = 1; i < static_cast<int>(prices.size()); ++i) {
            if (prices[i] < minPrice) {
                minPrice = prices[i];
            } else {
                best = max(best, prices[i] - minPrice);
            }
        }
        return best;
    }
};

File: array/problem121/problem121.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # Track the minimum price seen so far and best profit ending today
        min_price = prices[0]
        best = 0
        for p in prices[1:]:
            if p < min_price:
                min_price = p
            else:
                best = max(best, p - min_price)
        return best
